src/ngraph/pass/concat_fusion.cpp

```diff
@@ -0,0 +1,304 @@
+            c++;
+        }
+        ss << "-----------------------------" << std::endl;
+        return ss.str();
+    };
+
+    if (matcher->match(n))
+    {
+        auto concat_op = matcher->get_pattern_map()[concat_op_label];
+        if (!std::dynamic_pointer_cast<op::Concat>(concat_op))
+        {
+            NGRAPH_DEBUG << "self_concat_fusion: Pattern matcher matched incorrect op. Matched "
+                         << concat_op->get_name() << " instead of a self concat";
+            return;
+        }
+        if (!valid_self_concat(concat_op))
+        {
+            NGRAPH_DEBUG << "self_concat_fusion: " << concat_op->get_name()
+                         << " is not a valid self concat\n";
+            return;
+        }
+        else
+        {
+            NGRAPH_DEBUG << "self_concat_fusion: " << concat_op->get_name()
+                         << " is a VALID self concat\n";
+        }
+
+        auto& concat_vectors = this->m_concat_pattern_vectors;
+        size_t concat_axis =
+            std::static_pointer_cast<op::Concat>(concat_op)->get_concatenation_axis();
+        if (concat_vectors.empty())
+        {
+            concat_vectors.push_back(
+                make_pair(NodeVector{concat_op}, std::vector<size_t>{concat_axis}));
+
+            NGRAPH_DEBUG << print_state_of_bounded_vectors();
+        }
+        else
+        {
+            update_concat_pattern_vectors(concat_op, concat_axis);
+
+            NGRAPH_DEBUG << print_state_of_bounded_vectors();
+        }
+    }
+}
+
+void ngraph::pass::SelfConcatFusion::update_concat_pattern_vectors(
+    const std::shared_ptr<Node>& concat_op, size_t concat_axis)
+{
+    bool concat_source_found = false;
+    for (auto& concat_pattern_vec : this->m_concat_pattern_vectors)
+    {
+        auto last_op_in_pattern_vec = concat_pattern_vec.first.back();
+        NGRAPH_DEBUG << "self_concat_fusion: " << concat_op->get_name()
+                     << " trying to match source " << last_op_in_pattern_vec->get_name() << "\n";
+        if ((concat_op->get_argument(0) == last_op_in_pattern_vec) &&
+            (check_concat_has_no_fan_out(last_op_in_pattern_vec)))
+        {
+            NGRAPH_DEBUG << "MATCHED SOURCE " << concat_op->get_name() << " and "
+                         << last_op_in_pattern_vec->get_name() << "\n";
+            concat_pattern_vec.first.push_back(concat_op);
+            concat_pattern_vec.second.push_back(concat_axis);
+            concat_source_found = true;
+            break;
+        }
+        else
+        {
+            NGRAPH_DEBUG << "COULD NOT MATCH SOURCE " << concat_op->get_name() << " and "
+                         << last_op_in_pattern_vec->get_name() << "\n";
+        }
+    }
+
+    if (!concat_source_found)
+    {
+        this->m_concat_pattern_vectors.push_back(
+            make_pair(NodeVector{concat_op}, std::vector<size_t>{concat_axis}));
+    }
+}
+
+void ngraph::pass::SelfConcatFusion::remove_single_concat_op_pattern()
```

Comment:
Do we really need to remove single concat? What will happen if concat fusion work on a single concat?

Reply:
If I do not remove the single self concat, it will get replaced by a reshape and a broadcast. As you had mentioned once, that a single self concat is handled as a special case using arAssign, IMO replacing it with a reshape and a broadcast will be redundant and might even be slightly inefficient.

Reply:
Let me know if you think replacing a single self concat with reshape+broadcast would be equally/more efficient. I am not sure of that.

Reply:
reshape and broadcast is also implemented as `arAssign`. But since it is a ngraph pass I'm OK with what you have right now. Can simplify it later if necessary.